gc/base/EvacuatorScanspace.hpp

```diff
@@ -0,0 +1,252 @@
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#ifndef EVACUATORSCANSPACE_HPP_
+#define EVACUATORSCANSPACE_HPP_
+
+#include "Base.hpp"
+#include "EvacuatorCopyspace.hpp"
+#include "EvacuatorWhitelist.hpp"
+#include "IndexableObjectScanner.hpp"
+#include "ObjectScanner.hpp"
+#include "ObjectScannerState.hpp"
+
+/**
+ * Extends copyspace to allow scanning.
+ */
+class MM_EvacuatorScanspace : public MM_EvacuatorCopyspace
+{
+/*
+ * Data members
+ */
+private:
+	GC_ObjectScannerState _objectScannerState;	/* space reserved for instantiation of object scanner for current object */
+	GC_ObjectScanner *_objectScanner;			/* points to _objectScannerState after scanner is instantiated, NULL if object scanner not instantiated */
+	uint8_t *_scan;								/* scan head points to object being scanned */
+	uint8_t *_limit;							/* can copy object of size S in current stack frame as long as _copy<_limit and S<=(_end-_copy) */
+
+protected:
+	/* enumeration of flag bits may be extended past scanEndFlag by subclasses */
+	typedef enum scanspaceFlags {
+		isRememberedFlag = copyEndFlag << 1		/* remembered state of object at scan head */
+		, isSplitArrayFlag = copyEndFlag << 2	/* indicates that scanspace contains a slit array segment when set */
+		, scanEndFlag = isSplitArrayFlag		/* marks end of scanspace flags */
+	} scanspaceFlags;
+
+public:
+/*
+ * Function members
+ */
+private:
+
+protected:
+
+public:
+	/**
+	 * Basic array constructor obviates need for stdlibc++ linkage in gc component libraries. Array
+	 * is allocated from forge as contiguous block sized to contain requested number of elements and
+	 * must be freed using MM_Forge::free() when no longer needed.
+	 *
+	 * @param count the number of aray elements to instantiate
+	 * @return a pointer to instantiated array
+	 */
+	static MM_EvacuatorScanspace *
+	newInstanceArray(MM_Forge *forge, uintptr_t count)
+	{
+		MM_EvacuatorScanspace *scanspace = (MM_EvacuatorScanspace *)forge->allocate(sizeof(MM_EvacuatorScanspace) * count, OMR::GC::AllocationCategory::FIXED, OMR_GET_CALLSITE());
```

Comment:
Where is this memory freed?

Reply:
In MM_Evacuator::tearDown().

Reply:
Could this be placed in a comment for the newInstanceArray so it is clear to everyone where it is being freed.

Reply:
As above.